### DISCUSSION TITLE
[Lesson 32] [Fund subscription] Question about importing abstract contract containing constant variables

### DISCUSSION THREAD

**Discussion by tcrin** [Cyfrin/foundry-full-course-cu#2954] [Q&A] [answered]
Hello everyone,

I'm encountering an issue when trying to access the `LOCAL_CHAIN_ID` constant from an abstract contract in `HelperConfig.s.sol`. Even though `LOCAL_CHAIN_ID` is declared as a `public constant` in `CodeConstants`, I am unable to access this variable when importing the contract into `Interactions.s.sol`.

When I build the project, I get the following error:
```bash
~/foundry-f23/foundry-smart-contract-lottery$ forge build
[⠊] Compiling...
[⠰] Compiling 45 files with Solc 0.8.19
[⠔] Solc 0.8.19 finished in 253.68ms
Error: 
Compiler run failed:
Error (7576): Undeclared identifier.
  --> script/Interactions.s.sol:54:30:
   |
54 |         if (block.chainid == LOCAL_CHAIN_ID) {
   |                              ^^^^^^^^^^^^^^
```
**What I have tried:**

- **Checked the visibility** of the `LOCAL_CHAIN_ID` variable (it is `public constant`).
- **Imported the file correctly**.
- **Hardcoded the chain ID value**, which works, but I would prefer a more flexible solution.

❓ ❓Can anyone help me understand why the `LOCAL_CHAIN_ID` variable is not accessible in `Interactions.s.sol`?
 I would greatly appreciate any insights or suggestions!

---

#### `HelperConfig.s.sol`
```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import {Script} from "forge-std/Script.sol";
import {VRFCoordinatorV2_5Mock} from "@chainlink/contracts/src/v0.8/vrf/mocks/VRFCoordinatorV2_5Mock.sol";
import {LinkToken} from "../test/mocks/LinkToken.sol";

abstract contract CodeConstants {
    uint256 public constant SEPOLIA_ETH_CHAIN_ID = 1115511;
    uint256 public constant LOCAL_CHAIN_ID = 31337;
}

contract HelperConfig is CodeConstants, Script {
    // Rest of my code
}
```

#### `Interactions.s.sol`
```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import {Script} from "forge-std/Script.sol";
import {HelperConfig, CodeConstants} from "./HelperConfig.s.sol";
import {console} from "forge-std/console.sol";
import {VRFCoordinatorV2_5Mock} from "@chainlink/contracts/src/v0.8/vrf/mocks/VRFCoordinatorV2_5Mock.sol";
import {LinkToken} from "../test/mocks/LinkToken.sol";

contract FundSubscription is Script {
    uint256 public constant FUND_AMOUNT = 3 ether; // 3 LINK

    function fundSubscription(address vrfCoordinator, uint256 subscriptionId, address linkTokenAddress) public {
        console.log("Funding subscription: ", subscriptionId);
        console.log("On chainId: ", block.chainid);

        if (block.chainid == LOCAL_CHAIN_ID) { // This line causes build fail
            // My code here
        }
    }

    function run() external {
        fundSubscriptionUsingConfig();
    }
}
```

Thanks in advance!

**Comment by NikolayPIvanov** [comment, marked as answer]
From what I know so far, this has to do with how constants are stored in Solidity - at compile-time they become part of the bytecode - [reference to Solidity docs](https://docs.soliditylang.org/en/latest/contracts.html#constant-and-immutable-state-variables)

This means that in your example, `LOCAL_CHAIN_ID` which is part of  `CodeConstants` contract is nowhere in the bytecode of the `FundSubscription` contract. I think that if you inherit `CodeConstants` your code will compile

```solidity
contract FundSubscription is Script, CodeConstants {
  ...
}
```

If you wish to have something that you do not need to inherit, check out [libraries](https://docs.soliditylang.org/en/latest/contracts.html#libraries).

**Reply by tcrin** [reply]
Thank you very much! I’ve inherited `CodeConstants` and the build was successful. Your answer was extremely helpful and has given me a better understanding of how constants are stored in Solidity.

**Reply by EngrPips** [reply]
Awesome.

**Comment by heheboii11** [comment]
Hey @tcrin, you can see that in the error you have provided it states that ```LOCAL_CHAIN_ID```  as undeclared. Please pull the```LOCAL_CHAIN_ID``` from the helperconfig as below
```solidity
uint256 local_chain_id = HelperConfig.LOCAL_CHAIN_ID();```
and also you need to deploy the helper config in ```interactions.s.sol```, so the code might look like this 
```solidity 
contract FundSubscription is Script {
    uint256 public constant FUND_AMOUNT = 3 ether; // 3 LINK
    HelperConfig helper;
    function fundSubscription(address vrfCoordinator, uint256 subscriptionId, address linkTokenAddress) public {
        console.log("Funding subscription: ", subscriptionId);
        console.log("On chainId: ", block.chainid);
        helper  = new HelperConfig();
        uint256 local_chain_id = HelperConfig.LOCAL_CHAIN_ID();   
        if (block.chainid == local_chain_id) { // This line causes build fail // If you want you can do like this or just pass saving the variable in here in the interaction.s.sol, my advice is just store the variable in here
            // My code here
        }
    }

    function run() external {
        fundSubscriptionUsingConfig();
    }
}
```

**Reply by tcrin** [reply]
Thanks a lot! This is also a great suggestion, and I’ll give it a try. However, for this issue, I’ve already found the appropriate solution — simply inheriting `CodeConstants` works

**Reply by heheboii11** [reply]
Hey, thank you for bringing a new thing that I can learn, has no idea that we can inherit constant!!.

**Reply by tcrin** [reply]
Hey @heheboii11,
inherit constant - it seems you are misunderstanding. 
The reason I can inherit CodeConstants is because it is an abstract contract

**Reply by EngrPips** [reply]
I think you can inherit non-abstract contracts too, Or am I not remembering correctly?

**Reply by tcrin** [reply]
That's correct, you can inherit contracts or abstract contracts. However, in the case @heheboii11 mentioned, inherit constants is not possible.

**Reply by EngrPips** [reply]
> That's correct, you can inherit contracts or abstract contracts. However, in the case @heheboii11 mentioned, inherit constants is not possible.

![Screenshot 2024-12-02 at 03 41 10](https://github.com/user-attachments/assets/65e94375-2277-4711-8838-32841e726101)
